BGL: Alpha Expansion Graphcut

sgiraudot opened the pull request:
## Rationale

The Package _Surface Mesh Segmentation_ contains, internally, 3 implementations of the Alpha Expansion Graphcut algorithm: 2 using Boost, 1 using MaxFlow (third party software shipped with CGAL, under GPL version). This Graphcut is already used by _Surface Mesh Segmentation_ and _Classification_, and we have plans with @sloriot to use it in BGL too for regularizing mesh selections.

This Small Feature moves the algorithm to the BGL package, modifies the API so that it's BGL-like and documents it.

The algorithm is general enough to be applied to a Boost Graph concept and can benefit to many applications in geometry processing. In addition, this removes a weird dependency from Classification to Surface Mesh Segmentation.

## Summary of API changes

A new function `CGAL::alpha_expansion_graphcut()` is provided by the BGL package.

## License and copyright ownership

- Ownership was already and remains to GeometryFactory
- The Boost versions are moved to BGL and thus become LGPL (they used to be GPL in Segmentation)
- The MaxFlow version remains in Segmentation and remains under GPL (the third party software forbids us to change this)

## CHANGES.md

```markdown
### CGAL and the Boost Graph Library (BGL)
- Added function `alpha_expansion_graphcut()` which regularizes a
  multi-label partition over a user-defined graph.
```

## Submission

* ~~[__New User Manual Section__](https://cgal.geometryfactory.com/~sgiraudot/BGL-Alpha_expansion_graphcut/BGL/index.html#title33)~~
* ~~[__Reference Manual of `CGAL::alpha_expansion_graphcut()`__](https://cgal.geometryfactory.com/~sgiraudot/BGL-Alpha_expansion_graphcut/BGL/group__PkgBGLPartition.html#ga542bd1cc8c12fc4f5ca08c5b9acdc8f5)~~
* [__New User Manual Section__ after reviews](https://cgal.geometryfactory.com/~sgiraudot/BGL-Alpha_expansion_graphcut_v2/BGL/index.html#title33)
* [__Reference Manual of `CGAL::alpha_expansion_graphcut()`__ after reviews](https://cgal.geometryfactory.com/~sgiraudot/BGL-Alpha_expansion_graphcut_v2/BGL/group__PkgBGLPartition.html#ga1f2918778289a06ea2902783a1b87e18)

## Status

- Developed and locally tested (GNU/Linux)
- [Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Alpha_Expansion_Graphcut) -- pre-approved Afabri 2020-03-11

lrineau commented on Surface_mesh_segmentation/include/CGAL/boost/graph/alpha_expansion_graphcut_maxflow.h:
```diff
@@ -0,0 +1,103 @@
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0+
```
License issue: the CGAL package `BGL` is supposed to be LGPLv3+.

Reported by Travis: https://travis-ci.org/CGAL/cgal/jobs/562464389#L2760

sgiraudot replied:
This file is in Surface Mesh Segmentation for this very reason (see the __License__ section of my small feature), I don't understand the Travis error.

lrineau commented on BGL/include/CGAL/boost/graph/alpha_expansion_graphcut.h:
```diff
@@ -0,0 +1,717 @@
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0+
```
This file is actually the culprit: its SPDX tag says LGPL-v3+ but the text of the license above says GPLv3+.

sgiraudot replied:
Aaah, okay, I only modified the name of the license and forgot to change the text… I'll fix it, thanks.

lrineau commented:
https://travis-ci.org/CGAL/cgal/jobs/563427837#L7645
```
[100%] Built target packages_dependencies
 Checks finished
 Checks finished
Differences in Classification: Surface_mesh_segmentation have disappeared.
```

afabri commented:
Please add an example.

sloriot commented:
* `VertexLabelMap`probably only requires to be writable
* It should be made obvious that all vectors in `VertexLabelCostMap` should have the same size which is the number of labels, and that each label is represented by an index from 0->n-1
* Regularizes -> regularizes
* optional named parameters should be used for the vertex index map and the tag.

sgiraudot commented:
>     * Please add an example.
Done (with the input/output shown in the manual for better understanding).
>     * graphcut -> graph cut
Done.
>     * "Optimal partition from a set of labels can be computed" -> "An optimal partition..."
Done.
>     * You write "a model of Graph", but do we have such a concept?
I replaced by `VertexAndEdgeListGraph` which is the correct concept, as we discussed.

>     * `VertexLabelMap`probably only requires to be writable
No, it's also readable. The algorithm regularizes an _existing_ partition.
>     * It should be made obvious that all vectors in `VertexLabelCostMap` should have the same size which is the number of labels, and that each label is represented by an index from 0->n-1
Done.
>     * Regularizes -> regularizes
Done.
>     * optional named parameters should be used for the vertex index map and the tag.
Done.

I'll update the new doc in the top post.

MaelRL commented:
@afabri  @sloriot A conclusion on the small feature round is required.

afabri commented:
User Manual 
- "which, or a graph"
- partition or partitioning ??
- Why three terms for somehow the same:  energy, weight, cost ?
- "provided by the Triangulated Surface Mesh Segmentation Reference"   it should come from BGL package

Reference Manual
- partitions???
- Doesn't the word "seek" include the possibility to not succeed?
- Why is the value type of `VertexLabelCostMap`  a `std::vector` ?

afabri commented:
User Manual 
- "which, or a graph"
- partition or partitioning ??
- Why three terms for somehow the same:  energy, weight, cost ?
- "provided by the Triangulated Surface Mesh Segmentation Reference"   it should come from BGL package

Reference Manual
- partitions???
- Doesn't the word "seek" include the possibility to not succeed?
- Why is the value type of `VertexLabelCostMap`  a `std::vector` ?

sgiraudot commented:
Thanks for the review, I corrected where needed.

>     * partition or partitioning ??

I replaced using "partition" everywhere.

>     * Why three terms for somehow the same:  energy, weight, cost ?

Indeed. I replaced everything by "cost" which is the clearest terms of all, and I think it makes the documentation easier to understand now.

>     * "provided by the Triangulated Surface Mesh Segmentation Reference"   it should come from BGL package

No, there's a problem of license, it's GPL whereas BGL is LGPL (see the `License` section of my small feature).

>     * Doesn't the word "seek" include the possibility to not succeed?

Replaced by "compute".

>     * Why is the value type of `VertexLabelCostMap`  a `std::vector` ?

It seemed the most indicated way to embed a sequence of costs for a number of labels unknown at compilation time. What else should I use?

sloriot commented:
Broken: https://github.com/CGAL/cgal-dev/commit/813242bf8436b1bb835d60034433f6b250066188 cc@MaelRL I tried the new mechanism but it does not work.

sloriot commented:
```
BGL/include/CGAL/boost/graph/alpha_expansion_graphcut.h:262:35: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
     if(get (vertex_label_map, vd) != alpha
```

sloriot commented:
```
include/CGAL/boost/graph/alpha_expansion_graphcut.h:591:57: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
         double sink_weight = (get(vertex_label_map, vd) == alpha ?
```

MaelRL commented:
Should get good now, for index-related issues.

MaelRL commented:
There's also a doc warning (/home/cgal-testsuite/cgal_doc_build/CGAL-5.1-Ic-106/doc/BGL/BGL.txt:678: warning: \cite command to 'boykov2001fastapproximate' does not have an associated number)

MaelRL commented:
And in Polyhedron: /alpha_expansion_graphcut.h:178:1: warning: 'Alpha_expansion_boost_adjacency_list_tag' defined as a class here but previously declared as a struct; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]

sloriot commented:
Would be could to have that plugged into the selection plugin. Would it be easy?

sgiraudot commented:
> Would be could to have that plugged into the selection plugin. Would it be easy?

I actually have another branch that implements the `regularize_border_selection()` function that uses the alpha expansion to regularize borders. Because the API was still discussed at the time I worked on it, I decided to do first a PR for the alpha expansion API (this one) and another one for the application to border regularization (todo).

The other branch (that I have to update with this one and to make a small feature from) has the option in the selection plugin.

maxGimeno commented:
warnings https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-106.shtml#BGL_graphcut_Examples

maxGimeno commented:
remaining warnings https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-107/BGL_graphcut_Examples/TestReport_cgaltester_x86-64_Darwin-18.2_Apple-clang-default_Release.gz

sgiraudot commented:
There's still [a warning](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-108/Surface_mesh_segmentation_Examples/TestReport_Friedrich_Ubuntu-gcc-7.gz) that I can't manage to fix. It seems to be related to named parameters, most probably the implementation tag… but I did exactly the same thing as for `DiagonalizeTraits` which works well and I don't really understand the warning I get here.

sloriot commented:
FYI, we identified a pretty good suspect so no one needs to further look at it for now.

maxGimeno commented:
remaining warnings : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-109.shtml#BGL_graphcut_Examples

MaelRL commented on BGL/doc/BGL/BGL.txt:
```diff
@@ -653,6 +653,75 @@ can be done directly using `Face_filtered_graph`.
+  license and provided by the \ref PkgSurfaceMeshSegmentationRef
+  package
+
+All these implementation produce the exact same result but behave
```
```suggestion
All these implementations produce the exact same result but behave
```

MaelRL commented on BGL/doc/BGL/BGL.txt:
```diff
@@ -653,6 +653,75 @@ can be done directly using `Face_filtered_graph`.
+
+\cgalFigureBegin{alpha_exp, alpha_expansion.png}
+Comparison of time and memory consumed by the different alpha
+expansion implementation.
```
```suggestion
expansion implementations.
```

maxGimeno commented:
Succesfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-112.shtml

sloriot commented:
Tested in CGAL-5.1-Ic-114
